### PR TITLE
[WIP] add unit test for dashboard/index.js and plugin/index.js

### PR DESCRIPTION
**Pull request by ElreyB**
@ryan-roemer @parkerziegler The test are testing some default and optional settings but I wanted to go back test the utils file since it seems to be the base fo a lot of other functionality. Also, wanted to give you some chance to see where I have been going with testing and see if it is a good direction to be heading in.  My search in how to test cli has not been as fruitful as I would hope but I think it can be done. I think the hardest part I am having is knowing what the data structure of arguments look like as they are pass to functions. In turn making hard to make mock data to test functions. I am not that familiar with mocha and mocking data as well. So in reviewing this PR any advice or ideas on how to tackle the following would be extremely useful...

- How to mock things in mocha.
- Are there ways to test Clis?
- Ways to see the data structures that are being pass in functions

### CODE REVIEW

**Comment by ryan-roemer** [comment]
@ElreyB -- You've got lint failures and your editor I think has changed more existing stuff than was intended. Try `yarn lint --fix` to fix all the autofixable stuff and show you errors for the rest.

File: test/plugin/index.spec.js [hunk 1]
```diff
@@ -1,10 +1,49 @@
...
-const Plugin = require("../../plugin");
+const Plugin = require('../../plugin');
+let plugin;
+let pluginWithOptions;
+const options = { port: 3000, host: '111.0.2.3' };
 
-describe("plugin", () => {
-  it("can create a new plugin", () => {
-    const plugin = new Plugin();
+beforeEach(() => {
```
Comment (ryan-roemer) [inline test/plugin/index.spec.js]: NIT: Scope the `plugin` + other vars _in_ the `describe` as well as this `beforeEach`. That way, we compartmentalize things better if someone greps.

**Comment by ElreyB** [comment]
@ryan-roemer @parkerziegler Added test to the some of the utils files. The other utils file either heavily use dependency and/or had complex data structures.

File: test/dashboard/index.spec.js [hunk 2]
```diff
@@ -5,8 +5,48 @@ require("../base.spec");
...
+  let dashboard;
+  let dashboardWithOptions;
+  const options = {
+    color: "red",
+    minimal: true,
+    title: "my-title"
+  };
+
+  beforeEach(() => {
+    dashboard = new Dashboard();
```
Comment (ryan-roemer) [inline test/dashboard/index.spec.js]: I'd actually remove this `beforeEach` and have just two specs:

```js
it("can create a new no option dashboard", () => {
  const dashboard = new Dashboard();
  // ... all the dashboard asserts.
});

it("can create a new with options dashboard", () => {
  const dashboard = new Dashboard(options);
  // ... all the dashboard with options asserts.
});
```

and combine all the asserts for like objects there...

File: test/plugin/index.spec.js [hunk 3]
```diff
@@ -3,8 +3,52 @@
...
+  let plugin;
+  let pluginWithOptions;
+  const options = {
+    port: 3000,
+    host: "111.0.2.3"
+  };
+
+  beforeEach(() => {
+    plugin = new Plugin();
+    pluginWithOptions = new Plugin(options);
```
Comment (ryan-roemer) [inline test/plugin/index.spec.js]: Same comment as for dashboards above.

File: test/utils/format-assets.spec.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      const assets = [{ size: 500 }, { size: undefined }, { size: 1000 }];
+      expect(_getTotalSize(assets)).to.equal("1.46 KB");
+    });
+  });
+
+  describe("#_printAssets", () => {
+    it("returns a nested array of assets information", () => {
+      const assetList = [
+        {
+          name: "assest1",
```
Comment (ryan-roemer) [inline test/utils/format-assets.spec.js]: GLOBAL NIT: `s/assest/assets/`

File: test/utils/format-output.spec.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    context("when message is a type error", () => {
+      it("returns false", () => {
+        const message = "Type error: null has no properties";
+        expect(_isLikelyASyntaxError(message)).to.be.false;
+      });
+    });
+  });
+
+  describe("#_formatMessage", () => {
+    const message1 = "Module build failed: SyntaxError: missing ; before statement";
```
Comment (ryan-roemer) [inline test/utils/format-output.spec.js]: Stick the `message*` variables _in_ the `it()`

File: test/dashboard/index.spec.js (hunk 2 above)
Comment (ElreyB) [reply test/dashboard/index.spec.js]: Being new to Cil testing I would love to hear your take on writing the test one way over the other?

**Comment by ElreyB** [comment]
@ryan-roemer I refactor the unit test. I left a comment/question about one of the changes. It could also be something we can talk about in passing at the office.

File: test/dashboard/index.spec.js (hunk 2 above)
Comment (ryan-roemer) [reply test/dashboard/index.spec.js]: I tend to look at it as the unit of a test is:

1. Setup
2. State / changes
3. Asserts

Here, we don't have state / changes, we just have (1) setup (`new Dashboard()`) and asserts `expect()`. I think the best approach is combine these. So, for example, could look like for these tests are:

```js
describe("dashboard", () => {
  it("can create a new no option dashboard", () => {
    const dashboard = new Dashboard();

    expect(dashboard).to.be.ok;
    expect(dashboard.color).to.equal("green");
    expect(dashboard.minimal).to.be.false;
    expect(dashboard.stats).to.be.null;
  });

  it("can create a new with options dashboard", () => {
    const dashboardWithOptions = new Dashboard(options);

    expect(dashboardWithOptions.color).to.equal("red");
    expect(dashboardWithOptions.minimal).to.be.true;
  });
});
```

One other thing, is maybe use the `property` assert for better failure messages like:

```js
describe("dashboard", () => {
  it("can create a new no option dashboard", () => {
    const dashboard = new Dashboard();

    expect(dashboard).to.be.ok;
    expect(dashboard).to.have.property("color", "green");
    expect(dashboard).to.have.property("minimal", false);
    expect(dashboard).to.have.property("stats", null);
  });

  it("can create a new with options dashboard", () => {
    const dashboardWithOptions = new Dashboard(options);

    expect(dashboardWithOptions).to.be.ok;
    expect(dashboardWithOptions).to.have.property("color", "red");
    expect(dashboardWithOptions).to.have.property("minimal", true);
  });
});
```

Comment (ElreyB) [reply test/dashboard/index.spec.js]: Cool beans thank you.

**Comment by parkerziegler** [comment]
Will review tonight and merge! Thanks @ElreyB and @ryan-roemer, awesome work!

**Comment by parkerziegler** [comment]
All looks good on my end, tests pass locally and on CI. Going to merge it now, thanks you two!

**Pull request merged**